enhancements/dns/forward-dns-over-tls.md

```diff
@@ -0,0 +1,318 @@
+DNS request forwarding improves the current security posture for DNS on 
+OpenShift.
+
+### Goals
+
+1. A cluster administrator can specify a CA certificate or CA bundle, TLS configuration, and an upstream TLS Server Name for the default zone in the DNS configuration.
+2. A cluster administrator can specify a CA certificate or CA bundle, TLS configuration, and an upstream TLS Server Name for a non-default zone.
+3. A cluster administrator can opt-in to DNS-over-TLS (using the configured CA/bundle) for each listed upstream in the default zone.
+4. A cluster administrator can opt-in to DNS-over-TLS (using the configured CA/bundle) for a set of upstreams on a non-default zone.
+
+### Non-Goals
+
+This enhancement applies only to the CoreDNS Operator shipped as part of OpenShift. Anything outside that scope would be not applicable to this enhancement including the CoreDNS static pods managed by the `machine-config-operator`.
+
+## Proposal
+
+This proposal covers changes to two APIs. The first is a change to 
+`UpstreamResolvers`.  To enable a cluster administrator to configure 
+DNS-over-TLS, the `UpstreamResolvers` type will get two new optional members:
+
+`CABundle` references a ConfigMap that contains a single CA certificate 
+or a CA bundle.
+
+`ServerName` contains an FQDN to validate when establishing the connection. 
+This is relevant in particular with cases where a customer has multiple 
+upstreams behind one serving certificate (e.g. `1.1.1.1` and `2.2.2.2` 
+behind `dns.myorg.com`) as well as indicating what server we want to connect to.
+
+Since `UpstreamResolvers` doesn't allow for configuration of a custom zone, 
+it's assumed that the same TLS configuration will apply to _n_ `Upstream` 
+objects. 
+
+```go
+type UpstreamResolvers struct {
+  // ...<snip>
+  
+  // caBundle references a ConfigMap that must contain either a single
+  // CA Certificate or a CA Bundle (in the case of multiple upstreams signed
+  // by different CAs). Populating a configmap name in this field 
+  // effectively enables DNS-over-TLS. If this is empty or the configmap 
+  // referenced does not exist, DNS-over-TLS will not be enabled.
+  //
+  // 1. The configmap must contain a `cabundle.crt` key 
+  // 2. The value must be a PEM encoded certificate or certificate bundle.
+  // 3. The administrator must create this configmap in the openshift-config namespace.
+  // 4. The ServerName value must be supplied when CABundle is configured.
+  //
+  // +optional
+  // +kubebuilder:validation:Optional
+  CABundle corev1.ConfigMapNameReference `json:caBundle,omitempty`
+  
+  // serverName is the upstream server to connect to for DNS resolution. 
+  // This is required when a CABundle is supplied. Populating the 
+  // ServerName without a CABundle configured will have no effect.
+  //
+  // +optional
+  // +kubebuilder:validation:Pattern=^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$
```

Comment:
Each label (the portion between dots) can be up to 63, not 61, and a hyphen is allowed in any label.